pkg/gather/gather.go

```diff
@@ -0,0 +1,254 @@
+	ctx context.Context,
+	gatherer Interface,
+	rec recorder.Interface,
+	configurator configobserver.Configurator,
+	anonymizer *anonymization.Anonymizer,
+) error {
+	resultsChan, err := StartGatheringConcurrently(ctx, gatherer, configurator.Config().Gather)
+	if err != nil {
+		return err
+	}
+
+	gathererName := gatherer.GetName()
+
+	var errs []error
+	var functionReports []GathererFunctionReport
+
+	for result := range resultsChan {
+		if result.Panic != nil {
+			errStr := fmt.Sprintf(
+				"gatherer %v's function %v panicked with error: %v",
+				gathererName, result.FunctionName, result.Panic,
+			)
+
+			if result.IgnoreErrors {
+				klog.Error(errStr)
+			} else {
+				panic(errStr)
+			}
+		}
+
+		for _, err := range result.Errs {
+			errStr := fmt.Sprintf(
+				"gatherer %v's function %v failed with error: %v",
+				gathererName, result.FunctionName, err,
+			)
+
+			if result.IgnoreErrors {
+				klog.Error(errStr)
+			} else {
+				errs = append(errs, fmt.Errorf(errStr))
+			}
+		}
+
+		for _, r := range result.Records {
+			if err := rec.Record(r); err != nil {
+				errs = append(errs, fmt.Errorf(
+					"unable to record gatherer %v function %v' result %v because of error: %v",
+					gathererName, result.FunctionName, r.Name, err,
+				))
+			}
+		}
+
+		klog.Infof(
+			"Gather %v's function %v took %v to process %v records",
+			gathererName, result.FunctionName, result.TimeElapsed, len(result.Records),
+		)
+
+		functionReports = append(functionReports, GathererFunctionReport{
+			FuncName:     result.FunctionName,
+			Duration:     result.TimeElapsed,
+			RecordsCount: len(result.Records),
+			Errors:       result.Errs,
+			Panic:        result.Panic,
+		})
+	}
+
+	gathererMetadata := record.Record{
+		Name: fmt.Sprintf(gathererMetadataFilenamePattern, gathererName),
+		Item: record.JSONMarshaller{Object: GathererStatusReport{
+			StatusReports: functionReports,
+		}},
+	}
+	if err := rec.Record(gathererMetadata); err != nil {
+		errs = append(errs, fmt.Errorf("unable to record gatherer metadata because of error: %v", err))
+	}
+
+	var m runtime.MemStats
+	runtime.ReadMemStats(&m)
+
+	// if both gatherers get to the same archive we will get only the latest archive-metadata, which is totally fine
+	archiveMetadata := record.Record{
+		Name: archiveMetadataFilename,
+		Item: record.JSONMarshaller{Object: ArchiveMetadata{
+			MemoryAlloc:                m.HeapAlloc,
+			Uptime:                     time.Since(programStartTime).Truncate(time.Millisecond).Seconds(),
+			IsGlobalObfuscationEnabled: anonymizer != nil,
+		}},
+	}
+	if err := rec.Record(archiveMetadata); err != nil {
+		errs = append(errs, fmt.Errorf("unable to record archive metadata because of error: %v", err))
+	}
+
+	return sumErrors(errs)
+}
+
+// StartGatheringConcurrently starts gathering of enabled functions of the provided gatherer and returns a channel
+// with results which will be closed when processing is done
+func StartGatheringConcurrently(
```

Comment:
Does this function have to be exported (starting with upper case letter)?

Reply:
Actually no, it can be private, it will be used only in this package and in unit tests